### DISCUSSION TITLE
Next.js 13 app directory dynamic page

### DISCUSSION THREAD

**Discussion by igor-pestov-dev** [vercel/next.js#54887] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.14.2
      npm: 8.5.0
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 13.4.7
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: N/A
```

### Which example does this report relate to?

dynamic-routing, with-redux-toolkit

### What browser are you using? (if relevant)

Version 116.0.5845.140 (Official Build) (arm64)

### How are you deploying your application? (if relevant)

next build / next start, next dev

### Describe the Bug

```
// profile Page
import React from 'react';
import { cookies, headers } from 'next/headers';
import Profile from './profile';
import store from '../../../redux/store';
import getCommonHeaders from '../../../utils/getCommonHeaders';
import getUser from '../../../utils/user/getUser';
import { getPageProfile } from '../../../redux/reducers/pageProfile';

async function Page({ params }) {
  const user = getUser(cookies());
  const headersReq = getCommonHeaders(headers(), user, true);

  await store.dispatch(getPageProfile({
    headers: headersReq,
    userUrl: params.userUrl,
  }));
  const data = await store.getState().pageProfile;

  if (data.error) {
  // Here i want to set the status of the page when i call notFound from next/navigation i get 404 but the request may return 
  //another error as well
    throw new Error(data.error);
  }

  return (
    <Profile
      {...data}
      userUrl={params.userUrl}
      user={user}
    />
  );
}

// Redux file
export const getPageProfile = createAsyncThunk(
  'pageProfile/getPageProfile',
  async ({ headers, userUrl }) => {
    const response = await fetch(`${apiUrl}/${apiVersion}/page-profile/${userUrl}`, {
      headers,
    });
    if (!response.ok) throw new Error(response.status);

    const data = await response.json();

    return data;
  },
);
export const pageProfileSlice = createSlice({
  name: 'pageProfile',
  initialState: {
    notification: {},
    loading: false,
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getPageProfile.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(getPageProfile.fulfilled, (state, action) => {
        state.loading = false;
        state.data = action.payload.data;
      })
      .addCase(getPageProfile.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      });
  },
});
```
<img width="1154" alt="Screenshot 2023-09-01 at 14 18 48" src="https://github.com/vercel/next.js/assets/104328560/6024aef6-5bf6-425b-b9fd-a4ba8d764ad4">

### Expected Behavior

I would like to be able to change the status of the page, how can I do it with notFound, but I would like to set a status that will return my request, not just 404

### To Reproduce

I believe that the information provided gives a complete understanding of what I wanted to receive.

**Comment by eben-vranken** [comment]
I believe currently you are only throwing an error if the response status is not 200, you can set custom status codes by changing your API response

**`code`:**
```
export const getPageProfile = createAsyncThunk(
  'pageProfile/getPageProfile',
  async ({ headers, userUrl }) => {
    const response = await fetch(`${apiUrl}/${apiVersion}/page-profile/${userUrl}`, {
      headers,
    });

    if (!response.ok) {
      // Customize the error handling here
      if (response.status === 404) {
        throw new Error('Resource not found');
      } else if (response.status === 500) {
        throw new Error('Internal server error');
      } else {
        throw new Error('Other error');
      }
    }

    const data = await response.json();

    return data;
  },
);
```

**Reply by igor-pestov-dev** [reply]
@eben-vranken, !response.ok - only fires if the status is not 200, in which case any other status is required (400, 404, 500, 503, etc.). Then it turns out that I have one status in error handling, and on the page it is 200.

**Comment by devjiwonchoi** [comment]
I've created a related request, no progress after all since we cannot directly modify the status codes after the header is sent.

#53225

**Reply by igor-pestov-dev** [reply]
If I use notFound instead of throw new Error() , the status will change to 404. It would be nice to do this with other statuses as well.

**Reply by devjiwonchoi** [reply]
`not-found` also returns 200 when the response is streamed.

> Next.js will return a 200 HTTP status code for streamed responses, and 404 for non-streamed responses.

### Reference
- [not-found.js](https://nextjs.org/docs/app/api-reference/file-conventions/not-found)
